### PR TITLE
feat: observable messaging

### PR DESCRIPTION
**Pull request by mkazlauskas**
# Context

Web extension based wallets run in multiple processes (background, different UIs etc.). Running a separate instance of wallet in each process is very inefficient.

# Proposed Solution

The concept of sharing API objects remotely as implemented for cip30 is a nice solution. Extend the utilities to support exposing observables ([2aadcb7](https://github.com/input-output-hk/cardano-js-sdk/pull/244/commits/2aadcb768e9dd44ac1593e922f256dacf5225507)) and nested api object properties ([6593bba](https://github.com/input-output-hk/cardano-js-sdk/pull/244/commits/6593bbad481dd33776fe93d91033cadc1fdee432)) which should cover all that's needed to expose SingleAddressWallet.

# Important Changes Introduced

-  refactor: hoist testScheduler from wallet package [a42d79e](https://github.com/input-output-hk/cardano-js-sdk/pull/244/commits/a42d79e4a8b24f7d4e229a55b677b74af2f78f5b) & [722ca2f](https://github.com/input-output-hk/cardano-js-sdk/pull/244/commits/722ca2f58d903b937784d9d4420b0d44def31d83). note: wrong type in 1st commit - will update when squashing
- refactor!: hoist cip30 mapping of ObservableWallet to cip30 pkg [5dbe76d](https://github.com/input-output-hk/cardano-js-sdk/pull/244/commits/5dbe76d036b172238ce679dfcbc4ff238ea7c061)
- feat!: require to explicitly specify exposed api property names (for security reasons) and refactor remote api method options to be specified per method instead of per api object [4d8e446](https://github.com/input-output-hk/cardano-js-sdk/pull/244/commits/4d8e446081e7109c2186d73f76d49425f93f542b) & [ae2bcab](https://github.com/input-output-hk/cardano-js-sdk/pull/244/commits/ae2bcab5fabf555d1686b7c1bb1d31d51cca3191).
- Slightly refactor ObservableWallet interface so that it's all async and doesn't expose some unnecessary pieces [6742848](https://github.com/input-output-hk/cardano-js-sdk/pull/244/commits/6742848299f729451fe202681b6691ba324cd68b)

# Follow-up work

- Create a new `util` package with observable extensions and other general utils (not cardano-specific). Move TrackerSubject and Shutdown interface from `wallet` package and review utils in `core` package to probably hoist some of those as well.
- Fix test commented out in [02564bb](https://github.com/input-output-hk/cardano-js-sdk/pull/244/commits/02564bb47cccbba598173aa85f5d4fcb0b8e0489). It causes a timeout error in CI (could be a rxjs timeout error, not jest timeout). These tests pass locally, so functionality it covers should be good.

### CODE REVIEW

File: packages/web-extension/src/messaging/remoteApi.ts [hunk 1]
```diff
@@ -3,59 +3,133 @@
...
+  isObservable,
+  map,
+  merge,
+  mergeMap,
+  take,
+  takeUntil,
+  tap,
+  timeout
+} from 'rxjs';
+// Review: this import doesn't feel right - might make sense to hoist this to a shared util package.
```
[the remark this answers is not in the document]
Comment (mkazlauskas) [reply packages/web-extension/src/messaging/remoteApi.ts:32]: Agree. Some utils from core should also go there, such as `serializableObject`, `isNotNil` etc. Can we capture it as a separate task?

[the remark this answers is not in the document]
Comment (mkazlauskas) [reply packages/web-extension/src/messaging/remoteApi.ts]: Removed the comment in [a3861c1](https://github.com/input-output-hk/cardano-js-sdk/pull/244/commits/a3861c1f9187bf010ce7c87f1f17d062d3243383)

[the remark this answers is not in the document]
Comment (rhyslbw) [reply packages/web-extension/src/messaging/remoteApi.ts]: ADP-1772

**Review by rhyslbw — APPROVED** [review]
Great work @mkazlauskas

**Pull request merged**